lib/std/io/auto_indenting_stream.zig

```diff
@@ -0,0 +1,150 @@
+const std = @import("../std.zig");
+const io = std.io;
+const mem = std.mem;
+const Allocator = mem.Allocator;
+const ArrayList = std.ArrayList;
+const assert = std.debug.assert;
+
+/// Automatically inserts indentation of written data by keeping
+/// track of the current indentation level
+pub fn AutoIndentingStream(comptime indent_delta: usize, comptime WriterType: type) type {
+    return struct {
+        const Self = @This();
+        pub const Error = WriterType.Error;
+        pub const PushError = Allocator.Error;
+        pub const Writer = io.Writer(*Self, Error, write);
+        const Stack = ArrayList(usize);
+
+        writer_pointer: *WriterType,
+        indent_stack: Stack,
```

Comment:
This abstraction introduces heap allocation where it really shouldn't be necessary. This is a regression in terms of resource utilization, and likely also performance. I think you can keep this same abstraction, but if you take away the need to support variable indentation widths, you can remove the need for heap allocation like this. As far as I'm aware, 100% of the API users of this abstraction don't need variable indentation widths.

Reply:
I *almost* agree 100%, there's just the one place that needs a different indentation and that's the two space indentation in the assembly block. I'll have a think of a way of supporting it without the need of tracking the stack.

Reply:
> two space indentation in the assembly block

We can change the rules on this if it means that `zig fmt` becomes fundamentally simpler to implement. Feel free to do that.

Reply:
Done some basic performance testing, running fmt on std takes 3.5secs before the change and 4.2secs after. With a few modifications I found that all of that lost time can be gotten back by replacing the indent stack with a counter. So definitely worth fixing.

Reply:
I managed to keep to 2 space indent setting the indentation level to 2 when rendering the inline asm elements. Not the *most* elegant solution but it works without compromising the current style.